src/wmtk/PolyMesh.hpp

```diff
@@ -0,0 +1,264 @@
+     */
+    Tuple split_face(const Tuple& t, const Tuple& s);
+
+    /**
+     * @brief Join the two faces across the edge of t
+     * 
+     * Requires the edge of t to be adjacent to two distinct faces and have consistent orientation
+     */
+    Tuple join_face(const Tuple& t, const Tuple& s);
+
+    /**
+     * @brief Split the common vertex of t and s with the new edge between the merge of their
+     * respective faces
+     * 
+     * Requires t and s to share a vertex but not a face and have consistent orientation
+     */
+    Tuple split_vertex(const Tuple& t, const Tuple& s);
+
+    /**
+     * @brief Join the two endpoints the edge of t
+     * 
+     * Requires at least one endpoint of h to have valence > 1, and if vertices are distinct cannot
+     * be both boundary, unless the edge is boundary,
+     */
+    Tuple join_vertex(const Tuple& t);
+
+    /**
+     * @brief Mark a face as a hole in the surface.
+     */
+    Tuple make_hole(const Tuple& t);
+
+    /**
+     * @brief Mark a face as not a hole in the surface.
+     */
+    Tuple fill_hole(const Tuple& t);
+
+    // TODO These only make sense for triangle meshes, but we do want them in the interface
+    // since they can be implemented using splice and bubble operations
+    Tuple split_edge(const Tuple& t, Accessor<long>& hash_accessor) override;
+    Tuple collapse_edge(const Tuple& t, Accessor<long>& hash_accessor) override;
+    Tuple swap_edge(const Tuple& t);
+
+    Tuple switch_tuple(const Tuple& tuple, PrimitiveType type) const override;
+
+    /**
+     * @brief Jump to the next halfedge by performing a switch of vertex and edge
+     */
+    Tuple next_halfedge(const Tuple& tuple) const;
+
+    /**
+     * @brief Jump to the previous halfedge by performing a switch of edge and vertex
+     */
+    Tuple prev_halfedge(const Tuple& tuple) const;
+
+    /**
+     * @brief Jump to the opposite halfedge by performing a switch of face and vertex
+     */
+    Tuple opp_halfedge(const Tuple& tuple) const;
+
+    bool is_connectivity_valid() const override;
+
+    Tuple tuple_from_id(const PrimitiveType type, const long gid) const override;
+
+    bool is_ccw(const Tuple& tuple) const override;
+    bool is_boundary(const Tuple& tuple) const override;
+    bool is_boundary_vertex(const Tuple& tuple) const override;
+    bool is_boundary_edge(const Tuple& tuple) const override;
+
+    bool is_valid(const Tuple& tuple, ConstAccessor<long>& hash_accessor) const override;
+
+    void initialize(
+        Eigen::Ref<const VectorXl> next,
+        Eigen::Ref<const VectorXl> opp,
+        Eigen::Ref<const VectorXl> bnd_loops);
+    void initialize(Eigen::Ref<const RowVectors3l> F);
+
+protected:
+    long id(const Tuple& tuple, PrimitiveType type) const override;
+
+private:
+    /**
+     * @brief Generate a new vertex handle
+     * 
+     * TODO: Check if should be implemented using request_simplex_indices
+     */
+    long new_vertex();
```

Comment:
in principle we never want to expose interfaces that use indices - even "internally"